pkg/metrics/metrics.go

```diff
@@ -0,0 +1,69 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+//Package metrics implements custom metrics for CAPC
+package metrics
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	"regexp"
+	crtlmetrics "sigs.k8s.io/controller-runtime/pkg/metrics"
+)
+
+// AcsCustomMetrics encapsulates all CloudStack custom metrics defined for the controller.
+type AcsCustomMetrics struct {
+	acsReconciliationErrorCount *prometheus.CounterVec
+	errorCodeRegexp             *regexp.Regexp
+}
+
+// NewCustomMetrics constructs an AcsCustomMetrics with all desired CloudStack custom metrics and any supporting resources.
+func NewCustomMetrics() AcsCustomMetrics {
+	customMetrics := AcsCustomMetrics{}
+	customMetrics.acsReconciliationErrorCount = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "acs_reconciliation_errors",
+			Help: "Count of reconciliation errors caused by ACS issues, bucketed by error code",
+		},
+		[]string{"acs_error_code"},
+	)
+	if err := crtlmetrics.Registry.Register(customMetrics.acsReconciliationErrorCount); err != nil {
+		if are, ok := err.(prometheus.AlreadyRegisteredError); ok {
```

Comment:
Maybe this registers the count with a central record?

Reply:
Yes, Register adds the new counter to the app's Prometheus registry.  The code here (and below) detects if this is the case, and if so returns that already registered one instead of a new one (the creation of which failed).